include/oneapi/dpl/experimental/kt/internal/esimd_radix_sort_dispatchers.h

```diff
@@ -0,0 +1,327 @@
+            __cgh.host_task([=]() { sycl::free(__mem, __q); });
+        };
+        return __m_q.submit(__dealloc_task);
+    }
+};
+
+template <typename _KernelName, bool __is_ascending, ::std::uint8_t __radix_bits, ::std::uint16_t __data_per_work_item,
+          ::std::uint16_t __work_group_size, typename _RngPack>
+sycl::event
+__onesweep(sycl::queue __q, _RngPack&& __pack, ::std::size_t __n)
+{
+    using _KeyT = typename _RngPack::_KeyT;
+    using _ValT = typename _RngPack::_ValT;
+    constexpr bool __has_values = _RngPack::__has_values;
+
+    using _EsimdRadixSortHistogram = oneapi::dpl::__par_backend_hetero::__internal::__kernel_name_provider<
+        __esimd_radix_sort_onesweep_histogram<_KernelName>>;
+    using _EsimdRadixSortScan = oneapi::dpl::__par_backend_hetero::__internal::__kernel_name_provider<
+        __esimd_radix_sort_onesweep_scan<_KernelName>>;
+    using _EsimdRadixSortSweepEven = oneapi::dpl::__par_backend_hetero::__internal::__kernel_name_provider<
+        ::std::conditional_t<__has_values, __esimd_radix_sort_onesweep_even_by_key<_KernelName>,
+                                           __esimd_radix_sort_onesweep_even<_KernelName>>>;
+    using _EsimdRadixSortSweepOdd = oneapi::dpl::__par_backend_hetero::__internal::__kernel_name_provider<
+        ::std::conditional_t<__has_values, __esimd_radix_sort_onesweep_odd_by_key<_KernelName>,
+                                           __esimd_radix_sort_onesweep_odd<_KernelName>>>;
+    using _EsimdRadixSortCopyback = oneapi::dpl::__par_backend_hetero::__internal::__kernel_name_provider<
+        ::std::conditional_t<__has_values, __esimd_radix_sort_onesweep_copyback_by_key<_KernelName>,
+                                           __esimd_radix_sort_onesweep_copyback<_KernelName>>>;
+
+    using _GlobalHistT = ::std::uint32_t;
+    constexpr ::std::uint32_t __bin_count = 1 << __radix_bits;
+
+    const ::std::uint32_t __sweep_work_group_count = oneapi::dpl::__internal::__dpl_ceiling_div(__n, __work_group_size * __data_per_work_item);
+    constexpr ::std::uint32_t __bit_count = sizeof(_KeyT) * 8;
+    constexpr ::std::uint32_t __stage_count = oneapi::dpl::__internal::__dpl_ceiling_div(__bit_count, __radix_bits);
+
+    constexpr ::std::uint32_t __global_hist_item_count = __bin_count * __stage_count;
+    const ::std::uint32_t __group_hist_item_count = __bin_count * __stage_count * __sweep_work_group_count;
+
+    // Memory is not going to be allocated for void value type
+    // TODO: make this more explicit to reduce coupling between __onesweep_memory_holder and __rng_pack
+    __onesweep_memory_holder<_GlobalHistT, _KeyT, _ValT> __mem_holder(__q);
+    __mem_holder.__keys_alloc_count(__n);
+    if constexpr (__has_values)
+    {
+        __mem_holder.__vals_alloc_count(__n);
+    }
+    __mem_holder.__global_hist_item_alloc_count(__global_hist_item_count);
+    __mem_holder.__group_hist_item_alloc_count(__group_hist_item_count);
+    __mem_holder.__allocate();
```

Comment:
I feel like these statements belong together. Do we need them separately or could we allocate the memory in the ctor?

Reply:
It makes the code slightly more complex. This is what I've got: [mem_holder_ctor.patch](https://github.com/oneapi-src/oneDPL/files/13569940/mem_holder_ctor.patch)
Two constructors are needed to avoid calling `sizeof(_ValueT)` when `_ValueT` is `void`. This might be changed, but I need to thinkg about how to do it the proper way. Perhaps, the whole design `__onesweep_memory_holder` should be reconsidered.

By the way, the current implementation is inspired by https://refactoring.guru/design-patterns/builder.

Reply:
I may be mistaken but I think you should be able to use an `enable_if` call checking the type of `_ValueT` to separate the two constructors at compile time based on the provided types without requiring a different number of arguments.  I think that should contain the information you need to know if it is key sort or pair sort.
For me, the important thing would be unifying the constructor call in your patch which is currently separated by a constexpr.


If we want to leave the current structure, it can at least be improved by renaming the functions to `__set_...`, to make it more clear what the individual functions are doing.

Reply:
I have noted it in #1311 to address later.